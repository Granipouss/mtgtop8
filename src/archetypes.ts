import { getPage } from './utils';

export type ShortDeck = {
  id: number;
  eventId: number;
  name: string;
};

export type Archetype = { id: number; name: string; decks: ShortDeck[] };

const parseName = ($el: Cheerio): string =>
  $el
    .eq(1)
    .text()
    .trim()
    .replace(/ decks$/, '');

const parseDeck = ($el: Cheerio): ShortDeck => {
  const regexp = /^<a href="event\?e=(\d+)&amp;d=(\d+)&amp;f=MO">(.+)<\/a>$/;
  const match = $el.find('td').eq(1).html()?.match(regexp);
  if (!match) throw new Error(`Cannot parse deck: ${$el.text()}`);

  const eventId = parseInt(match[1], 10);
  const id = parseInt(match[2], 10);
  const name = match[3].trim();

  return { eventId, id, name };
};

export const getArchetype = async (
  id: number,
  page = 1
): Promise<Archetype> => {
  const $ = await getPage(
    `https://www.mtgtop8.com/archetype?a=${id}&current_page=${page}`
  );

  const archetype: Archetype = {
    id,
    name: parseName($('.w_title')),
    decks: [],
  };

  $('table table[align=center] .hover_tr').each((_, el) => {
    archetype.decks.push(parseDeck($(el)));
  });

  return archetype;
};
