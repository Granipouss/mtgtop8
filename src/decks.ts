import { getPage } from './utils';

export type ShortArchetype = { name: string; id: number };

export type CardGroup = 'mainboard' | 'sideboard' | 'commander';

export type Card = { count: number; name: string; isCompanion?: boolean };

export type Deck = {
  id: number;
  eventId: number;
  format: string;
  name: string;
  archetype: ShortArchetype;
  player: string;
  result: string;
} & {
  [group in CardGroup]: Card[];
};

const parseFormat = ($el: Cheerio): string => {
  const match = $el.html()?.match(/(.+) <div/);
  if (!match) throw new Error(`Cannot parse format: ${$el.text()}`);
  return match[1].trim();
};

const parseArchetype = ($el: Cheerio): ShortArchetype => {
  const regexp = /^<a href="archetype\?a=(\d+)">(.+)<br>decks<\/a>$/;
  const match = $el.html()?.match(regexp);
  if (!match) throw new Error(`Cannot parse archetype: ${$el.text()}`);

  const id = parseInt(match[1], 10);
  const name = match[2].trim();

  return { id, name };
};

const parseCard = ($el: Cheerio): Card => {
  const text = $el.text();
  const match = text.match(/^(\d+) (.+)$/);
  if (!match) throw new Error(`Cannot parse card: ${text}`);

  const count = parseInt(match[1], 10);
  const name = match[2].trim();
  const isCompanion = $el.find('[id="companion_%%%"]').length > 0;

  const card: Card = { count, name };
  if (isCompanion) card.isCompanion = true;

  return card;
};

const parseGroup = ($el: Cheerio): CardGroup => {
  switch ($el.text()) {
    case 'COMMANDER':
      return 'commander';
    case 'SIDEBOARD':
      return 'sideboard';
    default:
      return 'mainboard';
  }
};

export const getDeck = async (eventId: number, id: number): Promise<Deck> => {
  const $ = await getPage(`https://www.mtgtop8.com/event?e=${eventId}&d=${id}`);

  if ($.html().includes('No event could be found.')) {
    throw new Error(`No deck ${id} for event ${eventId}`);
  }

  const deck: Deck = {
    id,
    eventId,
    name: $('table .chosen_tr :not([align=right]) .topic').text().trim(),
    player: $('table .chosen_tr [align=right] .topic').text().trim(),
    result: $('table .chosen_tr [align=center]').text().trim(),
    format: parseFormat($('table table td[align="center"]').first()),
    archetype: parseArchetype($('.Nav_link').eq(1)),
    mainboard: [],
    sideboard: [],
    commander: [],
  };

  let currentGroup: CardGroup = 'mainboard';
  $('table table table td td').each((_, el) => {
    const isCard = $('.hover_tr', el).length > 0;
    if (isCard) {
      deck[currentGroup].push(parseCard($(el)));
    } else {
      currentGroup = parseGroup($(el));
    }
  });

  return deck;
};
