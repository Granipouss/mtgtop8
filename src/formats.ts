import { getPage } from './utils';

export type Format = { name: string; code: string };

export const getFormats = async (): Promise<Format[]> => {
  const $ = await getPage(`https://www.mtgtop8.com/`);

  const formats: Format[] = [];

  $('table')
    .eq(1)
    .find('td')
    .each((_, el) => {
      const link = $(el).find('a').attr('href');
      const linkRegexp = /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/;
      const linkMatch = link?.match(linkRegexp);
      if (!linkMatch) return;
      const code = linkMatch[1].trim().toUpperCase();

      const alt = $(el).find('img').attr('id');
      if (!alt) return;
      const name = alt
        .replace(/^menu_/, '')
        .replace(/_/g, ' ')
        .toLowerCase();

      formats.push({ name, code });
    });

  $('#other_tooltip_content div').each((_, el) => {
    const link = $(el).find('a').attr('href');
    const linkRegexp = /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/;
    const linkMatch = link?.match(linkRegexp);
    if (!linkMatch) return;
    const code = linkMatch[1].trim().toUpperCase();

    const alt = $(el).find('img').attr('id');
    if (!alt) return;
    const name = alt
      .replace(/^menu_/, '')
      .replace(/_/g, ' ')
      .toLowerCase();

    formats.push({ name, code });
  });

  return formats;
};
