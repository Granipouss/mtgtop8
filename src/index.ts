export * from './formats';
export * from './decks';
export * from './archetypes';
