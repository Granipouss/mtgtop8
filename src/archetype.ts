import { getPage, url, match } from './utils';

import { ShortDeck } from './deck';

export type ShortArchetype = { id: number; name: string };

export class Archetype implements ShortArchetype {
  public static async get(id: number, page = 1): Promise<Archetype> {
    const $ = await getPage(url.archetype(id, page));
    return new Archetype($);
  }

  id: number;
  name: string;

  decks: ShortDeck[];

  private constructor($: CheerioStatic) {
    this.id = parseInt(match($('table table option').attr('value'), /archetype\?a=(\d+)&f=.+&meta=.+/)[0], 10);

    this.name = $('table table').eq(1).find('tr:first-child').text().replace('decks', '').trim();

    this.decks = $('table table[align=center] .hover_tr')
      .get()
      .map((el) => {
        const [eventString, deckString, name] = match(
          $(el).find('td').eq(1).html() || '',
          /^<a href="event\?e=(\d+)&amp;d=(\d+)&amp;f=MO">(.+)<\/a>$/
        );

        return {
          eventId: parseInt(eventString, 10),
          id: parseInt(deckString, 10),
          name: name.trim(),
        };
      });
  }
}
