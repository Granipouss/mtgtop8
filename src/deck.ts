import { getPage, url, match, ListOf, WithCount, parseDeckLink } from './utils';

import { ShortArchetype } from './archetype';
import { ShortFormat } from './format';

export type ShortDeck = { eventId: number; id: number; name: string };

export type CardGroup = 'mainboard' | 'sideboard' | 'commander';

export type Card = { name: string; isCompanion?: boolean };

export class Deck implements ShortDeck {
  public static async get(eventId: number, id: number): Promise<Deck> {
    const $ = await getPage(url.deck(eventId, id));

    // FIXME: check for deck also
    if ($.html().includes('No event could be found.')) {
      throw new Error(`No deck ${id} for event ${eventId}`);
    }

    return new Deck($);
  }

  eventId: number;
  id: number;
  name: string;

  format: ShortFormat;
  archetype: ShortArchetype;
  player: string;
  result: string;

  mainboard: ListOf<Card> = [];
  sideboard: ListOf<Card> = [];
  commander: ListOf<Card> = [];

  private constructor($: CheerioStatic) {
    const chosenLinkData = parseDeckLink($('.chosen_tr a').attr('href'));

    this.eventId = chosenLinkData.event;

    this.id = chosenLinkData.deck;

    this.name = $('table .chosen_tr :not([align=right]) .topic').text().trim();

    this.format = {
      name: match($('table table td[align="center"]').html() || '', /(.+) <div/)[0].trim(),
      code: chosenLinkData.format,
    };

    const [archetypeId, archetypeName] = match(
      $('.Nav_link').eq(1).html() || '',
      /^<a href="archetype\?a=(\d+)">(.+)<br>decks<\/a>$/
    );

    this.archetype = {
      id: parseInt(archetypeId, 10),
      name: archetypeName.trim(),
    };

    this.player = $('table .chosen_tr [align=right] .topic').text().trim();

    this.result = $('table .chosen_tr [align=center]').text().trim();

    let currentGroup: CardGroup = 'mainboard';
    $('table table table td td').each((_, el) => {
      const isCard = $('.hover_tr', el).length > 0;
      if (isCard) {
        this[currentGroup].push(parseCard($(el)));
      } else {
        currentGroup = parseGroup($(el));
      }
    });
  }
}

const parseCard = ($el: Cheerio): WithCount<Card> => {
  const text = $el.text();
  const match = text.match(/^(\d+) (.+)$/);
  if (!match) throw new Error(`Cannot parse card: ${text}`);

  const count = parseInt(match[1], 10);
  const name = match[2].trim();
  const isCompanion = $el.find('[id="companion_%%%"]').length > 0;

  const card: WithCount<Card> = { count, name };
  if (isCompanion) card.isCompanion = true;

  return card;
};

const parseGroup = ($el: Cheerio): CardGroup => {
  switch ($el.text()) {
    case 'COMMANDER':
      return 'commander';
    case 'SIDEBOARD':
      return 'sideboard';
    default:
      return 'mainboard';
  }
};
