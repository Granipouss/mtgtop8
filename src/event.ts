import { getPage, url, match, parseDeckLink, isDefined } from './utils';
import { ShortFormat } from './format';
import { ShortDeck } from './deck';

export type ShortEvent = { id: number; name: string };

type WithRanks<T> = T & { ranks: number[] };

export class Event implements ShortEvent {
  public static async get(id: number): Promise<Event> {
    const $ = await getPage(url.event(id));

    if ($.html().includes('No event could be found.')) {
      throw new Error(`No event ${id}`);
    }

    return new Event($);
  }

  id: number;
  name: string;

  format: ShortFormat;
  rank: number;
  playerCount: number;
  date: Date;
  decks: WithRanks<ShortDeck>[] = [];

  private constructor($: CheerioStatic) {
    const chosenLinkData = parseDeckLink($('.chosen_tr a').attr('href'));

    this.id = chosenLinkData.event;

    this.name = $('.w_title td:first-child').text().trim();

    this.format = {
      name: match($('table table td[align="center"]').html() || '', /(.+) <div/)[0].trim(),
      code: chosenLinkData.format,
    };

    const stars = $('img[src="/graph/star.png"]').length;
    const bigstars = $('img[src="/graph/bigstar.png"]').length;
    this.rank = 10 * bigstars + stars;

    const [playerCount, date] = match($('table table td[align="center"]').html() || '', /(\d+) players - (.+)/);

    this.playerCount = parseInt(playerCount, 10);

    this.date = new Date(`${date}Z`);

    this.decks = $('.hover_tr, .chosen_tr')
      .get()
      .map((el) => {
        try {
          const linkData = parseDeckLink($(el).find('a').first().attr('href'));
          return {
            ranks: $(el).find('div:first-child').text().split('-').map(Number),
            eventId: linkData.event,
            id: linkData.deck,
            name: $(el).find('a').first().text(),
          };
        } catch (error) {
          return;
        }
      })
      .filter(isDefined);
  }
}
