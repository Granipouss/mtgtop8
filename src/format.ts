import { getPage, match, ListOf, isDefined, url } from './utils';

import { ShortArchetype } from './archetype';
import { ShortEvent } from './event';

export type ShortFormat = { name: string; code: string };

export class Format implements ShortFormat {
  public static async get(code: string): Promise<Format> {
    const $ = await getPage(url.format(code));

    if (getTableRows($, /Metagame Breakdown/).length === 1) {
      throw new Error(`No format for code ${code}`);
    }

    return new Format($);
  }

  public static async getAll(): Promise<ShortFormat[]> {
    const $ = await getPage(url.base());
    return [...$('table').eq(1).find('td').get(), ...$('#other_tooltip_content div').get()]
      .map((el) => {
        try {
          const [code] = match($(el).find('a').attr('href'), /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/);

          const alt = $(el).find('img').attr('id');
          if (!alt) return;
          const name = alt
            .replace(/^menu_/, '')
            .replace(/_/g, ' ')
            .toLowerCase();

          return { name, code };
        } catch (error) {
          return;
        }
      })
      .filter(isDefined);
  }

  public name: string;
  public code: string;

  public archetypes: ListOf<ShortArchetype> = [];

  private constructor($: CheerioStatic) {
    this.code = match($('table table .hover_tr a').attr('href'), /archetype\?a=.+&meta=.+&f=(.+)/)[0];

    this.name = $('table table a[target=_blank]')
      .text()
      .replace(/ Construction Rules$/, '')
      .toLowerCase()
      .trim();

    this.archetypes = getTableRows($, /Metagame Breakdown/)
      .filter((el) => $(el).is('.hover_tr'))
      .map((el) => {
        const [idString] = match($(el).find('a').attr('href'), /archetype\?a=(\d+)&meta=.+&f=.+/);
        return {
          id: parseInt(idString, 10),
          name: $(el).find('a').text().trim(),
          count: parseInt($(el).find('[align=right]').first().text(), 10),
        };
      });
  }

  async getEvents(page = 1): Promise<ShortEvent[]> {
    const $ = await getPage(url.format(this.code, page));
    return getTableRows($, /^(Events \d+ to \d+|Last \d+ events)$/).map((el) => {
      const [idString] = match($(el).find('a').attr('href'), /event\?e=(\d+)&f=.+/);
      return {
        id: parseInt(idString, 10),
        name: $(el).find('a').text().trim(),
      };
    });
  }
}

const getTableRows = ($: CheerioStatic, title: RegExp): CheerioElement[] => {
  try {
    const table: CheerioElement | undefined = $('table table')
      .get()
      .find((el) => $('tr:first-child', el).text().match(title));
    return table ? $(table).find('tr').get().splice(1) : [];
  } catch (error) {
    return [];
  }
};
