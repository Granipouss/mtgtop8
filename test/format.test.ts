import { Format } from '../src';

describe('Format', () => {
  it('should get all the list', async () => {
    const formats = await Format.getAll();
    expect(formats).toEqual([
      { code: 'VI', name: 'vintage' },
      { code: 'LE', name: 'legacy' },
      { code: 'MO', name: 'modern' },
      { code: 'PI', name: 'pioneer' },
      { code: 'ST', name: 'standard' },
      { code: 'EDH', name: 'commander' },
      { code: 'PAU', name: 'pauper' },
      { code: 'PEA', name: 'peasant' },
      { code: 'BL', name: 'block' },
      { code: 'EX', name: 'extended' },
      { code: 'HIGH', name: 'highlander' },
      { code: 'CHL', name: 'canadian highlander' },
    ]);
  });

  it('should get one', async () => {
    await expect(Format.get('MO')).resolves.toBeTruthy();
  });

  it('should get paginated events', async () => {
    const format = await Format.get('ST');
    await Promise.all([
      expect(format.getEvents(1)).resolves.toEqual(expect.any(Array)),
      expect(format.getEvents(2)).resolves.toEqual(expect.any(Array)),
    ]);
  });

  it('should throw when not found', async () => {
    await expect(Format.get('XX')).rejects.toThrow('No format for code XX');
  });
});
