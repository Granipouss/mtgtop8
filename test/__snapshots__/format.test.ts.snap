// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get all formats 1`] = `
Array [
  Object {
    "code": "VI",
    "name": "vintage",
  },
  Object {
    "code": "LE",
    "name": "legacy",
  },
  Object {
    "code": "MO",
    "name": "modern",
  },
  Object {
    "code": "PI",
    "name": "pioneer",
  },
  Object {
    "code": "ST",
    "name": "standard",
  },
  Object {
    "code": "EDH",
    "name": "commander",
  },
  Object {
    "code": "PAU",
    "name": "pauper",
  },
  Object {
    "code": "PEA",
    "name": "peasant",
  },
  Object {
    "code": "BL",
    "name": "block",
  },
  Object {
    "code": "EX",
    "name": "extended",
  },
  Object {
    "code": "HIGH",
    "name": "highlander",
  },
  Object {
    "code": "CHL",
    "name": "canadian highlander",
  },
]
`;

exports[`should get events for a format 1`] = `
Array [
  Object {
    "id": 26659,
    "name": "Event @ Hareruya (Japan)",
  },
  Object {
    "id": 26636,
    "name": "Deluxe Cup  @ Hareruya (Japan)",
  },
  Object {
    "id": 26658,
    "name": "Manatraders Series Modern 15K",
  },
  Object {
    "id": 26606,
    "name": "MTGO Modern League",
  },
  Object {
    "id": 26611,
    "name": "Four Tops @ Hareruya (Japan)",
  },
  Object {
    "id": 26619,
    "name": "MTGO Modern Preliminary",
  },
  Object {
    "id": 26672,
    "name": "FNM @ Hareruya (Japan)",
  },
  Object {
    "id": 26844,
    "name": "Modern Leaders #05 @ Magic Mage Masters (Switzerland)",
  },
  Object {
    "id": 26583,
    "name": "Weekday Event @ Hareruya (Japan)",
  },
  Object {
    "id": 26535,
    "name": "MTGO Modern Champs #12182511",
  },
]
`;

exports[`should get one format 1`] = `
Format {
  "archetypes": Array [
    Object {
      "count": 43,
      "id": 23,
      "name": "Canadian Threshold",
    },
    Object {
      "count": 29,
      "id": 213,
      "name": "UR Delver",
    },
    Object {
      "count": 25,
      "id": 18,
      "name": "Goblins",
    },
    Object {
      "count": 21,
      "id": 319,
      "name": "Death & Taxes",
    },
    Object {
      "count": 16,
      "id": 553,
      "name": "Eldrazi Aggro",
    },
    Object {
      "count": 15,
      "id": 898,
      "name": "Ninja",
    },
    Object {
      "count": 8,
      "id": 78,
      "name": "Dragon Stompy",
    },
    Object {
      "count": 7,
      "id": 19,
      "name": "Merfolk",
    },
    Object {
      "count": 6,
      "id": 915,
      "name": "Esper Vial",
    },
    Object {
      "count": 6,
      "id": 149,
      "name": "Delver (Other)",
    },
    Object {
      "count": 6,
      "id": 479,
      "name": "Grixis Delver",
    },
    Object {
      "count": 5,
      "id": 772,
      "name": "Death's Shadow",
    },
    Object {
      "count": 5,
      "id": 20,
      "name": "Burn",
    },
    Object {
      "count": 5,
      "id": 811,
      "name": "Arclight Phoenix",
    },
    Object {
      "count": 4,
      "id": 95,
      "name": "Slivers",
    },
    Object {
      "count": 4,
      "id": 128,
      "name": "Maverick",
    },
    Object {
      "count": 2,
      "id": 44,
      "name": "Bant Aggro",
    },
    Object {
      "count": 1,
      "id": 798,
      "name": "Humans",
    },
    Object {
      "count": 1,
      "id": 301,
      "name": "Jund",
    },
    Object {
      "count": 1,
      "id": 338,
      "name": "Patriot Aggro",
    },
    Object {
      "count": 1,
      "id": 133,
      "name": "Pikula",
    },
    Object {
      "count": 1,
      "id": 56,
      "name": "Other - Aggro",
    },
    Object {
      "count": 18,
      "id": 36,
      "name": "Lands",
    },
    Object {
      "count": 17,
      "id": 766,
      "name": "4c Control",
    },
    Object {
      "count": 13,
      "id": 883,
      "name": "Bant Control",
    },
    Object {
      "count": 10,
      "id": 24,
      "name": "Loam",
    },
    Object {
      "count": 7,
      "id": 473,
      "name": "Artifacts Blue",
    },
    Object {
      "count": 6,
      "id": 172,
      "name": "Stoneblade",
    },
    Object {
      "count": 5,
      "id": 34,
      "name": "Landstill",
    },
    Object {
      "count": 5,
      "id": 217,
      "name": "Nic Fit",
    },
    Object {
      "count": 4,
      "id": 242,
      "name": "UWx Control",
    },
    Object {
      "count": 2,
      "id": 79,
      "name": "The Rock (Junk)",
    },
    Object {
      "count": 2,
      "id": 86,
      "name": "BUG Midrange",
    },
    Object {
      "count": 1,
      "id": 85,
      "name": "Dreadstill",
    },
    Object {
      "count": 1,
      "id": 88,
      "name": "Pox",
    },
    Object {
      "count": 1,
      "id": 178,
      "name": "BUG Control",
    },
    Object {
      "count": 1,
      "id": 478,
      "name": "Grixis Control",
    },
    Object {
      "count": 6,
      "id": 61,
      "name": "Other - Control",
    },
    Object {
      "count": 32,
      "id": 555,
      "name": "Dark Depths",
    },
    Object {
      "count": 26,
      "id": 33,
      "name": "Show and Tell",
    },
    Object {
      "count": 17,
      "id": 32,
      "name": "Reanimator",
    },
    Object {
      "count": 15,
      "id": 35,
      "name": "Storm",
    },
    Object {
      "count": 13,
      "id": 38,
      "name": "Elves",
    },
    Object {
      "count": 13,
      "id": 376,
      "name": "Cloudpost Ramp",
    },
    Object {
      "count": 13,
      "id": 161,
      "name": "Doomsday",
    },
    Object {
      "count": 11,
      "id": 21,
      "name": "Dredge",
    },
    Object {
      "count": 10,
      "id": 82,
      "name": "Aluren",
    },
    Object {
      "count": 8,
      "id": 851,
      "name": "Hogaak",
    },
    Object {
      "count": 6,
      "id": 87,
      "name": "Painter",
    },
    Object {
      "count": 5,
      "id": 773,
      "name": "Bomberman",
    },
    Object {
      "count": 4,
      "id": 368,
      "name": "Food Griffin",
    },
    Object {
      "count": 3,
      "id": 83,
      "name": "Belcher",
    },
    Object {
      "count": 2,
      "id": 96,
      "name": "Cephalid Breakfast",
    },
    Object {
      "count": 2,
      "id": 98,
      "name": "Infect",
    },
    Object {
      "count": 1,
      "id": 563,
      "name": "Mono Red Combo",
    },
    Object {
      "count": 1,
      "id": 91,
      "name": "Enchantress",
    },
    Object {
      "count": 1,
      "id": 66,
      "name": "Other - Combo",
    },
  ],
  "code": "LE",
  "name": "legacy",
}
`;
